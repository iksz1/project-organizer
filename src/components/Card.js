import React, { Component } from "react";
import PropTypes from "prop-types";
import style from "./Card.scss";
import { observer } from "mobx-react";
import EditForm from "./EditForm";
import EditIcon from "react-icons/lib/md/mode-edit";
import DeleteIcon from "react-icons/lib/md/delete";

@observer
class Card extends Component {
  static propTypes = {
    card: PropTypes.object,
    onCardDelete: PropTypes.func
  };

  state = {
    editMode: false
  };

  toggleEditMode = () => {
    this.setState(prevState => ({ editMode: !prevState.editMode }));
  };

  handleDoubleClick = () => {
    this.props.card.changeText("DOUBLE CLICKED!");
  };

  handleUpdate = text => {
    this.props.card.changeText(text);
    this.toggleEditMode();
  };

  render() {
    const { card, onCardDelete } = this.props;
    const editMode = this.state.editMode;

    if (editMode) {
      return (
        <EditForm text={card.text} onSubmit={this.handleUpdate} onCancel={this.toggleEditMode} />
      );
    }

    return (
      <div className={style.card} onDoubleClick={this.handleDoubleClick}>
        {card.text}
        <div className={`${style.controls} dont-drag`}>
          <button onClick={this.toggleEditMode}>
            <EditIcon />
          </button>
          <button onClick={() => onCardDelete(card)}>
            <DeleteIcon />
          </button>
        </div>
      </div>
    );
  }
}

export default Card;
