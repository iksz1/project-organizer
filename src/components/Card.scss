@import "../vars";

.card {
  padding: 0.5em 1em;
  margin-bottom: 0.5em;
  background: $bg;
  box-shadow: 0 1px 2px 0 rgba(34, 36, 38, 0.15);
  &:first-child {
    margin-top: 0.5em;
  }
}
