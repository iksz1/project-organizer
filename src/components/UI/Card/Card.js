import React, { Component } from "react";
import PropTypes from "prop-types";
import { observer } from "mobx-react";
import styled from "styled-components";
import EditForm from "../EditForm/EditForm";
import PopupButtons from "../PopupButtons/PopupButtons";

export const Wrapper = styled.div`
  padding: 0.5em 1em;
  margin-bottom: 0.5em;
  position: relative;
  background: ${props => props.theme.bgCard};
  border-radius: 0.2em;
  box-shadow: ${props => props.theme.boxShadowAlt};
  word-wrap: break-word;
`;

class Card extends Component {
  static propTypes = {
    card: PropTypes.object,
    onCardDelete: PropTypes.func
  };

  state = {
    editMode: false
  };

  toggleEditMode = () => {
    this.setState(prevState => ({ editMode: !prevState.editMode }));
  };

  handleUpdate = text => {
    this.props.card.changeText(text);
    this.toggleEditMode();
  };

  render() {
    const { card, onCardDelete } = this.props;
    const editMode = this.state.editMode;
    const popupItems = [
      { handler: this.toggleEditMode, icon: "edit" },
      { handler: () => onCardDelete(card), icon: "trash" }
    ];

    if (editMode) {
      return (
        <Wrapper>
          <EditForm text={card.text} onSubmit={this.handleUpdate} onCancel={this.toggleEditMode} />
        </Wrapper>
      );
    }

    return (
      <Wrapper className="draggable with-popup">
        {card.text}
        <PopupButtons items={popupItems} />
      </Wrapper>
    );
  }
}

export default observer(Card);
