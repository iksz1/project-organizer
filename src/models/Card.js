import { types, flow } from "mobx-state-tree";
import db from "../utils/dbWrapper";

export const Card = types
  .model("Card", {
    id: types.number,
    listId: types.number,
    boardId: types.number,
    text: types.string,
    order: types.optional(types.number, 0),
    labelId: types.optional(types.number, 0),
    deleted: types.optional(types.boolean, false)
  })
  .actions(self => ({
    changeText: flow(function* changeText(text) {
      yield db.update("cards", { id: self.id, text });
      self.text = text;
    })
  }));
