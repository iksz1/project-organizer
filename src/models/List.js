import { types, flow, destroy } from "mobx-state-tree";
import { Card } from "./Card";
import db from "../utils/dbWrapper";
import calcOrder from "../utils/reorderHelper";

export const List = types
  .model("List", {
    id: types.number,
    boardId: types.number,
    name: types.string,
    cards: types.optional(types.array(Card), []),
    order: types.optional(types.number, 0),
    deleted: types.optional(types.boolean, false)
  })
  .actions(self => ({
    changeName: flow(function* changeName(name) {
      yield db.update("lists", { id: self.id, name });
      self.name = name;
    }),

    addCard: flow(function* addCard(text) {
      const card = yield db.add("cards", {
        text,
        listId: self.id,
        boardId: self.boardId,
        order: calcOrder({ toArr: self.cards })
      });
      self.cards.push(card);
    }),

    deleteCard: flow(function* deleteCard(item) {
      yield db.remove("cards", item.id);
      destroy(item);
    })
  }));
