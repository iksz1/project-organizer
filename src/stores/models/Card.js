import { types, flow } from "mobx-state-tree";
import db from "../../utils/dbWrapper";

export const CardBase = types.model("CardBase", {
  id: types.number,
  listId: types.number,
  boardId: types.number,
  text: types.string,
  order: types.optional(types.number, 0),
  labelId: types.optional(types.number, 0),
  deleted: types.optional(types.number, 0)
});

const Card = CardBase.named("Card").actions(self => ({
  changeText: flow(function* changeText(text) {
    yield db.update("cards", { id: self.id, text });
    self.text = text;
  })
}));

export default Card;
