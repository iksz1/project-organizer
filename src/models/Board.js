import { types, flow, destroy, getParent } from "mobx-state-tree";
import { List } from "./List";
import db from "../utils/dbWrapper";
import calcOrder from "../utils/reorderHelper";

export const Board = types
  .model("Board", {
    id: types.number,
    name: types.string,
    lists: types.optional(types.array(List), []),
    order: types.optional(types.number, 0),
    deleted: types.optional(types.number, 0)
  })
  .actions(self => ({
    changeName: flow(function* changeName(name) {
      yield db.update("boards", { id: self.id, name });
      self.name = name;
    }),

    addList: flow(function* addList(name) {
      const list = yield db.add("lists", {
        boardId: self.id,
        name,
        order: calcOrder({ toArr: self.lists })
      });
      self.lists.push(list);
    }),

    deleteList: flow(function* deleteList(list) {
      yield db.remove("lists", list);
      destroy(list);
    }),

    moveList: flow(function* moveList(listToMove, meta) {
      const { fromIndex, toArr, toIndex } = meta;
      if (fromIndex === toIndex) return; //exit if position hasn't changed
      const order = calcOrder(meta);
      const list = { ...listToMove, order, cards: listToMove.cards.map(card => ({ ...card })) };
      yield db.update("lists", { id: list.id, order });
      // const lists = toArr.slice();
      toArr.splice(fromIndex, 1);
      toArr.splice(toIndex, 0, list);
      // lists.splice(toIndex, 0, lists.splice(fromIndex, 1)[0]);
      // lists[toIndex] = list;
      // toArr.replace(lists);
    }),

    moveCard: flow(function* moveCard(cardToMove, meta) {
      const { fromArr, fromIndex, toArr, toIndex } = meta;
      if (fromArr === toArr && fromIndex === toIndex) return; //exit if position hasn't changed

      if (fromArr === toArr) {
        //same list scenario
        const order = calcOrder(meta);
        const card = { ...cardToMove, order };
        yield db.update("cards", card);
        const cards = toArr.slice();
        cards.splice(toIndex, 0, cards.splice(fromIndex, 1)[0]);
        cards[toIndex] = card;
        toArr.replace(cards);
      } else {
        //different lists scenario
        const order = calcOrder(meta);
        const card = { ...cardToMove, listId: getParent(toArr).id, order };
        yield db.update("cards", card);
        toArr.splice(toIndex, 0, card);
        fromArr.splice(fromIndex, 1);
      }
    })
  }));
