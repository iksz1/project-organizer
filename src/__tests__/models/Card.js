import db from "../../utils/dbWrapper";
import { Card } from "../../models/Card";

jest.mock("../../utils/dbWrapper");

const cardData = {
  id: 1,
  listId: 1,
  boardId: 1,
  text: "decorators are stage-2"
};

it("creates instance successfully", async () => {
  const card = Card.create(cardData);
  db.update.mockResolvedValue();
  await card.changeText("correct");
  expect(card.toJSON()).toMatchSnapshot();
});
