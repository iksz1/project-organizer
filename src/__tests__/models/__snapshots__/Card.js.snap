// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates instance successfully 1`] = `
Object {
  "boardId": 1,
  "deleted": 0,
  "id": 1,
  "labelId": 0,
  "listId": 1,
  "order": 0,
  "text": "correct",
}
`;
